pkg/querier/queryrange/limits.go

```diff
@@ -188,41 +190,123 @@ func (l limitsMiddleware) Do(ctx context.Context, r queryrangebase.Request) (que
 
 type querySizeLimiter struct {
 	next              queryrangebase.Handler
+	cfg               []config.PeriodConfig
 	maxQueryBytesRead func(string) int
 	errorTemplate     string
 }
 
 // NewQuerySizeLimiterMiddleware creates a new Middleware that enforces query size limits.
 // The errorTemplate should format two strings: the bytes that would be read and the bytes limit.
-func NewQuerySizeLimiterMiddleware(maxQueryBytesRead func(string) int, errorTemplate string) queryrangebase.Middleware {
+func NewQuerySizeLimiterMiddleware(
+	cfg []config.PeriodConfig,
+	maxQueryBytesRead func(string) int,
+	errorTemplate string,
+) queryrangebase.Middleware {
 	return queryrangebase.MiddlewareFunc(func(next queryrangebase.Handler) queryrangebase.Handler {
 		return &querySizeLimiter{
 			next:              next,
+			cfg:               cfg,
 			maxQueryBytesRead: maxQueryBytesRead,
 			errorTemplate:     errorTemplate,
 		}
 	})
 }
 
-// getIndexStatsForRequest return the index stats for the matchers in r's query
-func (q *querySizeLimiter) getIndexStatsForRequest(ctx context.Context, r queryrangebase.Request) (*logproto.IndexStatsResponse, error) {
-	matchers, err := syntax.ExtractMatchersFromQuery(r.GetQuery())
+// getBytesReadForRequest returns the number of bytes that would be read for the query in r.
+// Since the query expression may contain multiple stream matchers, this function sums up the
+// bytes that will be read for each stream.
+// E.g. for the following query:
+//
+//	count_over_time({job="foo"}[5m]) / count_over_time({job="bar"}[5m] offset 10m)
+//
+// this function will sum the bytes read for each of the following streams, taking into account
+// individual intervals and offsets
+//   - {job="foo"}
+//   - {job="bar"}
+//
+// TODO: This function shares a lot of code with dynamicShardResolver.Shards. I think we should extract the functionality to a common function?
+func (q *querySizeLimiter) getBytesReadForRequest(ctx context.Context, r queryrangebase.Request) (uint64, error) {
+	expr, err := syntax.ParseExpr(r.GetQuery())
 	if err != nil {
-		return nil, err
+		return 0, err
+	}
+
+	matcherGroups, err := syntax.MatcherGroups(expr)
+	if err != nil {
+		return 0, err
+	}
+
+	// TODO: Set concurrency dynamically as in shardResolverForConf?
+	const maxConcurrentIndexReq = 10
+	results := make([]*stats.Stats, len(matcherGroups), len(matcherGroups))
+	if err := concurrency.ForEachJob(ctx, len(matcherGroups), maxConcurrentIndexReq, func(ctx context.Context, i int) error {
+		matchers := syntax.MatchersString(matcherGroups[i].Matchers)
+		diff := matcherGroups[i].Interval + matcherGroups[i].Offset
+		adjustedFrom := model.Time(r.GetStart()).Add(-diff)
+		if matcherGroups[i].Interval == 0 {
+			// TODO: IIUC, this is needed for instant queries.
+			//       Should we use the MaxLookBackPeriod as done in shardResolverForConf?
+			adjustedFrom = adjustedFrom.Add(-100)
```

Comment:
Yes, `MaxLookBackPeriod` is what we want here (and I think it's a good idea to extract this to a utility somewhere we can reuse)

Reply:
Done with https://github.com/grafana/loki/pull/8670/commits/aaf90e5e0ce7ae497caa7b60b8dd346bbd81c283